README.md
<h1 align="center">曝光台</h1>
<p align="center">
基于 Web3 和 Vue.JS 的 dAPP
</p>

## 背景
区块链是基于 P2P 网络、共识机制和加密算法等技术的新型应用模式，有广阔的应用前景。该曝光台基于区块链，利用区块链去中心化、数据难以篡改的特点，提供更优越的业务功能。
### 优势
- 管理高效性  
    
    智能合约的灵活运用提高管理效率，曝光台的透明性、实时性有助于使用者的自动化、精细化管理，有效降低了人力成本，保证了沟通质量的稳定性。
- 系统安全性  
    
    - 在内容层面，区块链上的数据无法被篡改；
    - 在结构层面，区块链数据的分布式存储使得少量节点的故障难以影响整个区块链的安全；
    - 在访问层面，区块链技术的加密机制将没有相关密钥的节点拒之门外，降低了非法访问的可能性。
## 总体架构
- [智能合约](ET.sol)

    使用 solidity 语言，负责管理用户信息和曝光信息、控制数据操作权限，系统节点包含、数据管理服务。
- 以太坊客户端

    使用 go-ethereum 客户端，将系统节点连接到以太坊网络，从而调用合约函数发送交易（注册用户、创建曝光、回复曝光）。以太坊客户端会对用户提供的以太坊账户地址、密钥对进行检查实现身份认证。之后，用户从系统节点发出的所有交易都由该账户发起，并由该账户提供发起交易所需的以太币。
- [用户界面](UI/src/components/ET.vue)

    使用 Web3 实现用户界面和以太坊客户端的交互，使用 Vue.JS 搭建直观友好的操作界面，降低用户学习成本。
## 使用说明
### 前期准备
需要自行[安装 geth](https://github.com/ethereum/go-ethereum/wiki/Installing-Geth)。

然后克隆该仓库。
```bash
~ $ git clone https://github.com/Ernie1/exposure-table.git && cd exposure-table
```
### 启动以太坊客户端
新建多个位于 exposure-table 文件夹位置的终端窗口以启动多个节点。
```bash
exposure-table $ cd user0 && geth --datadir . console --identity "user0" --port "30303" --rpc --rpccorsdomain="*" --rpcport "8545" --rpcapi "db,eth,net,web3,miner,personal,admin" --networkid 15 console
```
```bash
exposure-table $ cd user1 && geth --datadir . console --identity "user1" --port "30304" --rpc --rpccorsdomain="*" --rpcport "8546" --rpcapi "db,eth,net,web3,miner,personal,admin" --networkid 15 console
```
```bash
exposure-table $ cd user2 && geth --datadir . console --identity "user2" --port "30305" --rpc --rpccorsdomain="*" --rpcport "8547" --rpcapi "db,eth,net,web3,miner,personal,admin" --networkid 15 console
```
```bash
exposure-table $ cd user3 && geth --datadir . console --identity "user3" --port "30306" --rpc --rpccorsdomain="*" --rpcport "8548" --rpcapi "db,eth,net,web3,miner,personal,admin" --networkid 15 console
```
如果需要添加更多的节点，每个节点需要在 exposure-table 文件夹下新建一个文件夹，然后初始化。
```bash
新的文件夹 $ geth --datadir . init ../genesis.json
```
并按照 user[0-3] 类似的方法启动节点，并获取节点信息。
```bash
> admin.nodeInfo
```
将返回结果的 `enode` 字段的值加入 [UI/config/index.js](UI/config/index.js) 的 `nodeInfo` 字段的数组中。
### 启动用户界面
```bash
exposure-table $ cd UI
UI $ npm install
UI $ npm start
```
打开浏览器访问 `http://localhost:8080` 。可以同时打开多个界面操作不同的节点。

#### 总体效果
![](screenshot/2018-12-17-19.25.55.png)
#### 操作流程
输入 rpc 端口号连接到对应的以太坊客户端。
![](screenshot/2018-12-17-14.45.57.png)
如果当前节点没有以太坊账户，需要设置一个密码新建账户。
![](screenshot/2018-12-16-18.01.42.png)
![](screenshot/2018-12-16-18.02.03.png)
![](screenshot/2018-12-16-18.02.10.png)
![](screenshot/2018-12-16-18.02.13.png)
如果当前节点没有合约（曝光台）用户，需要设置用户名和是否被监督来注册用户，勾选“被监督”的用户身份为监督对象，否则为监督者。
![](screenshot/2018-12-16-18.10.04.png)
![](screenshot/2018-12-16-18.11.19.png)
注册合约用户是一笔以太坊交易，需要以太坊账户的权限，如果权限已经超时，需要输入密码解锁。
![](screenshot/2018-12-16-18.19.43.png)
正确输入密码后，回到“注册合约用户”对话框，再次点击“确定”，以太坊客户端会开始挖矿。
![](screenshot/2018-12-16-18.20.41.png)
交易完成（注册成功）后会自动停止挖矿，进入曝光台界面。
![](screenshot/2018-12-17-09.01.43.png)
已经注册合约用户的节点，下次进入用户界面时只需要输入对应的 rpc 端口号连接到以太坊客户端后就可以直接进入曝光台界面。
![](screenshot/2018-12-17-13.31.08.png)
监督者用户可以点击左侧监督对象条目创建曝光。被监督者用户点击无效。
![](screenshot/2018-12-18-15.54.18.png)
![](screenshot/2018-12-17-15.33.31.png)
![](screenshot/2018-12-17-16.05.01.png)
创建曝光是一笔以太坊交易，需要以太坊账户的权限，如果权限已经超时，需要输入密码解锁。
![](screenshot/2018-12-17-15.34.15.png)
正确输入密码后，回到“创建曝光”对话框，再次点击“确定”，以太坊客户端会开始挖矿。
![](screenshot/2018-12-17-16.05.13.png)
交易完成（创建曝光）后会自动停止挖矿，进入曝光台界面。
![](screenshot/2018-12-17-16.05.15.png)
对应的曝光对象的用户界面会实时显示横幅通知。
![](screenshot/2018-12-17-16.05.17.png)
![](screenshot/2018-12-17-16.08.00.png)
被监督者用户可以点击右侧曝光相应条目回复曝光。回复过的和不属于自己的点击无效。监督者用户点击无效。
![](screenshot/2018-12-18-15.52.22.png)
![](screenshot/2018-12-17-18.57.38.png)
回复曝光是一笔以太坊交易，需要以太坊账户的权限，如果权限已经超时，需要输入密码解锁。
![](screenshot/2018-12-17-19.14.12.png)
正确输入密码后，回到“回复曝光”对话框，再次点击“确定”，以太坊客户端会开始挖矿。交易完成（回复曝光）后会自动停止挖矿，进入曝光台界面。
![](screenshot/2018-12-17-19.14.27.png)
对应的曝光者的用户界面会实时显示横幅通知。
![](screenshot/2018-12-17-19.14.30.png)
![](screenshot/2018-12-17-19.19.04.png)
## 测试
连接以太坊客户端失败提示。
![](screenshot/2018-12-18-15.51.58.png)
注册合约用户时用户名已存在提示。
![](screenshot/2018-12-17-19.23.58.png)
解锁节点账户时密码错误提示。
![](screenshot/2018-12-18-15.52.43.png)

user4/history

admin.nodeInfo

test/app.js
let candidates = {"Rama": "candidate-1", "Nick": "candidate-2", "Jose": "candidate-3"}

$(document).ready(function() {
	//初始化余额
	candidateNames = Object.keys(candidates);
 	for (var i = 0; i < candidateNames.length; i++) {
		let voteName = candidateNames[i];
  		totalVotesFor(voteName); 	
	}

 	//初始化事件
 	$(".vote-btn").click(function(){
 		//获取投票人名称
 		let voteName=$(this).prev().prev().text();
 		voteForCandidate(voteName);

 	});
});

function totalVotesFor(voteName) {
    $.get("/totalVotesFor?voteName=" + voteName, function(data) {
		if(data == "Error") {
            alert('提示', '500');
        } else {
			$("#"+candidates[voteName]).html(data);
		}
    });
}

function voteForCandidate(voteName) {
    $.get("/voteForCandidate?voteName=" + voteName, function(data) {
        if(data == "Error") {
            alert('提示', '500');
        } else {
           let div_id = candidates[voteName];
		   var vote_num = totalVotesFor(voteName);
		   $("#"+div_id).html(data);//.fadeIn(800);
		   $("#"+div_id);//.fadeOut(400);
        }
    });
}


test/index.js
var express = require('express');
var app = express();
var server = require('http').createServer(app);
var Web3 = require("web3");

app.use(express.static('.'));

let web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:8545"));
//这里导入你自己的ABI
var votingContract = web3.eth.contract([{"constant":false,"inputs":[{"name":"candidate","type":"bytes32"}],"name":"totalVotesFor","outputs":[{"name":"","type":"uint8"}],"payable":false,"type":"function","stateMutability":"nonpayable"},{"constant":false,"inputs":[{"name":"candidate","type":"bytes32"}],"name":"validCandidate","outputs":[{"name":"","type":"bool"}],"payable":false,"type":"function","stateMutability":"nonpayable"},{"constant":true,"inputs":[{"name":"","type":"bytes32"}],"name":"votesReceived","outputs":[{"name":"","type":"uint8"}],"payable":false,"type":"function","stateMutability":"view"},{"constant":true,"inputs":[{"name":"","type":"uint256"}],"name":"candidateList","outputs":[{"name":"","type":"bytes32"}],"payable":false,"type":"function","stateMutability":"view"},{"constant":false,"inputs":[{"name":"candidate","type":"bytes32"}],"name":"voteForCandidate","outputs":[],"payable":false,"type":"function","stateMutability":"nonpayable"},{"inputs":[{"name":"candidateNames","type":"bytes32[]"}],"payable":false,"type":"constructor","stateMutability":"nonpayable"}]);
//这里要替换成你自己的地址
let contractInstance = votingContract.at('0x130d22e81cea6ae5b8e3e60bb75c9a75d3e84b59');

app.get("/totalVotesFor", function(req, res) {
	var voteName = req.query.voteName;
	var vote_num=contractInstance.totalVotesFor.call(voteName).toString();
	console.log(vote_num);
	res.send(vote_num);
});

app.get("/voteForCandidate", function(req, res) {
	var voteName = req.query.voteName;
	contractInstance.voteForCandidate(voteName, {from: web3.eth.accounts[0]});
	var vote_num=contractInstance.totalVotesFor.call(voteName).toString();
	res.send(vote_num);
});

server.listen(3000);

// 控制台会输出以下信息
console.log('Server running at http://127.0.0.1:3000/index.html');


user4/main.py
from icrawler.builtin import BingImageCrawler, GoogleImageCrawler
import threading

keywords = [
	'ruler tear paper',
	'ruler cutting paper',
	'cutting paper',
	'Knife unpacking box',
	'teeth tear tape',
	'bite tape',
	'finger knife unpacking box',
	'nimble finger knife']

filters = dict(
    type='photo')


def goo(keyword):
    google_crawler = GoogleImageCrawler(
        feeder_threads=1,
        parser_threads=1,
        downloader_threads=4,
        storage={'root_dir': keyword+' google'})
    google_crawler.crawl(keyword=keyword, offset=0,
                         filters=filters, max_num=200, file_idx_offset=0)


def bing(keyword):
    bing_crawler = BingImageCrawler(downloader_threads=4, storage={
                                    'root_dir': keyword+' bing'})
    bing_crawler.crawl(keyword=keyword, filters=filters, offset=0, max_num=200)


def goobing(keyword):
    threading.Thread(target=goo, args=(keyword,)).start()
    threading.Thread(target=bing, args=(keyword,)).start()


for k in keywords:
    threading.Thread(target=goobing, args=(k,)).start()



UI/config/index.js
'use strict'
// Template version: 1.3.1
// see http://vuejs-templates.github.io/webpack for documentation.

const path = require('path')

module.exports = {
  dev: {

    // Paths
    assetsSubDirectory: 'static',
    assetsPublicPath: '/',
    proxyTable: {},

    // Various Dev Server settings
    host: 'localhost', // can be overwritten by process.env.HOST
    port: 8080, // can be overwritten by process.env.PORT, if port is in use, a free one will be determined
    autoOpenBrowser: false,
    errorOverlay: true,
    notifyOnErrors: true,
    poll: false, // https://webpack.js.org/configuration/dev-server/#devserver-watchoptions-


    /**
     * Source Maps
     */

    // https://webpack.js.org/configuration/devtool/#development
    devtool: 'cheap-module-eval-source-map',

    // If you have problems debugging vue-files in devtools,
    // set this to false - it *may* help
    // https://vue-loader.vuejs.org/en/options.html#cachebusting
    cacheBusting: true,

    cssSourceMap: true
  },

  build: {
    // Template for index.html
    index: path.resolve(__dirname, '../dist/index.html'),

    // Paths
    assetsRoot: path.resolve(__dirname, '../dist'),
    assetsSubDirectory: 'static',
    assetsPublicPath: '/',

    /**
     * Source Maps
     */

    productionSourceMap: true,
    // https://webpack.js.org/configuration/devtool/#production
    devtool: '#source-map',

    // Gzip off by default as many popular static hosts such as
    // Surge or Netlify already gzip all static assets for you.
    // Before setting to `true`, make sure to:
    // npm install --save-dev compression-webpack-plugin
    productionGzip: false,
    productionGzipExtensions: ['js', 'css'],

    // Run the build command with an extra argument to
    // View the bundle analyzer report after build finishes:
    // `npm run build --report`
    // Set to `true` or `false` to always turn it on or off
    bundleAnalyzerReport: process.env.npm_config_report
  },
  nodeInfo:
    [
      "enode://21e6344b3f21095796bc5a66cc709edef739eccc3b0448e38c3b8c17da2d72fb4b929131ac5ffce102560f1fd357ab7eccda897dd2eaa7570abca852ec1e5744@127.0.0.1:30303",
      "enode://e15c1ca927cdee306301edf431879795937dfa49fa5d007ef98fc6ab49f0a02ac041fcab807bd64112024c754fc97da04b5a2181238c1652979602938eeed562@127.0.0.1:30304",
      "enode://fc550a7c567ed0071611ee79f8dcf324840db00a3aefb58b132f3cff801cf30da679b961885b264921dc8b98c2a3c06cb077df87867ec6c80be6d00a975fdbf5@127.0.0.1:30305",
      "enode://c210aec5c712b995f8480eff1a7e8af5600b974481964fd25fe7807c030d52ac8351d521f6ccf705aea61269291f05eab0fbb79a1bd0f4ed0178238aa95beddf@127.0.0.1:30306"
    ],
  contractInterface: [{ "constant": true, "inputs": [{ "name": "i", "type": "uint256" }], "name": "exposureByIndex", "outputs": [{ "name": "", "type": "string" }, { "name": "", "type": "string" }, { "name": "", "type": "string" }, { "name": "", "type": "uint256" }, { "name": "", "type": "string" }, { "name": "", "type": "uint256" }], "payable": false, "stateMutability": "view", "type": "function" }, { "constant": true, "inputs": [{ "name": "i", "type": "uint256" }], "name": "superviseeByIndex", "outputs": [{ "name": "", "type": "string" }, { "name": "", "type": "uint256" }], "payable": false, "stateMutability": "view", "type": "function" }, { "constant": true, "inputs": [{ "name": "addr", "type": "address" }], "name": "useraddrExist", "outputs": [{ "name": "", "type": "bool" }], "payable": false, "stateMutability": "view", "type": "function" }, { "constant": true, "inputs": [], "name": "userInfo", "outputs": [{ "name": "", "type": "string" }, { "name": "", "type": "uint256" }, { "name": "", "type": "bool" }], "payable": false, "stateMutability": "view", "type": "function" }, { "constant": true, "inputs": [], "name": "exposureAmount", "outputs": [{ "name": "", "type": "uint256" }], "payable": false, "stateMutability": "view", "type": "function" }, { "constant": false, "inputs": [{ "name": "index", "type": "uint256" }, { "name": "reply", "type": "string" }], "name": "replyExposure", "outputs": [{ "name": "", "type": "bool" }], "payable": false, "stateMutability": "nonpayable", "type": "function" }, { "constant": true, "inputs": [{ "name": "name", "type": "string" }], "name": "usernameExist", "outputs": [{ "name": "", "type": "bool" }], "payable": false, "stateMutability": "view", "type": "function" }, { "constant": false, "inputs": [{ "name": "name", "type": "string" }, { "name": "supervisee", "type": "bool" }], "name": "userRegister", "outputs": [{ "name": "", "type": "bool" }], "payable": false, "stateMutability": "nonpayable", "type": "function" }, { "constant": false, "inputs": [{ "name": "superviseeName", "type": "string" }, { "name": "detail", "type": "string" }], "name": "createExposure", "outputs": [{ "name": "", "type": "bool" }], "payable": false, "stateMutability": "nonpayable", "type": "function" }, { "constant": true, "inputs": [{ "name": "username", "type": "string" }], "name": "exposureIndexesByName", "outputs": [{ "name": "", "type": "uint256[]" }, { "name": "", "type": "uint256" }], "payable": false, "stateMutability": "view", "type": "function" }, { "constant": true, "inputs": [], "name": "superviseeIndexes", "outputs": [{ "name": "", "type": "uint256[]" }, { "name": "", "type": "uint256" }], "payable": false, "stateMutability": "view", "type": "function" }, { "anonymous": false, "inputs": [{ "indexed": false, "name": "name", "type": "string" }, { "indexed": false, "name": "time", "type": "uint256" }], "name": "userRegisterEvent", "type": "event" }, { "anonymous": false, "inputs": [{ "indexed": false, "name": "index", "type": "uint256" }, { "indexed": false, "name": "superviserName", "type": "string" }, { "indexed": false, "name": "superviseeName", "type": "string" }, { "indexed": false, "name": "detail", "type": "string" }, { "indexed": false, "name": "time", "type": "uint256" }], "name": "createExposureEvent", "type": "event" }, { "anonymous": false, "inputs": [{ "indexed": false, "name": "index", "type": "uint256" }, { "indexed": false, "name": "reply", "type": "string" }, { "indexed": false, "name": "time", "type": "uint256" }], "name": "replyExposureEvent", "type": "event" }],
  contractAddress: '0x334d2c8ebcf9e06fb9ba87482e5857067cb4357c'
}
